### PR TITLE
Support token authentication

### PR DESCRIPTION
**Pull request by igor-vovk**
We are using token authentication in our consul setup. This pull request adds ability to sign all requests with token, if passed. I changed ConsulRequestBasics signature, moving from object to class, because I want to move building full uri from *Requests classes to ConsulRequestBasics in next PR.

### CODE REVIEW

File: src/main/scala/consul/v1/catalog/CatalogRequests.scala [hunk 1]
```diff
@@ -52,7 +51,9 @@ object CatalogRequests {
     Json.writes[Registerable]
   }
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): CatalogRequests = new CatalogRequests {
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): CatalogRequests = new CatalogRequests {
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/catalog/CatalogRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759076/issues/source?bid=2788492&fileBranchId=2939115#l56)

Comment (paulblei) [inline src/main/scala/consul/v1/catalog/CatalogRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759076/issues/source?bid=2788492&fileBranchId=2939115#l56)

File: src/main/scala/consul/v1/health/HealthRequests.scala [hunk 2]
```diff
@@ -18,7 +18,8 @@ object HealthRequests {
 
   implicit private val NodesHealthServiceReads: Reads[NodesHealthService] = Json.reads[NodesHealthService]
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): HealthRequests = new HealthRequests {
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): HealthRequests = new HealthRequests {
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/health/HealthRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759090/issues/source?bid=2788492&fileBranchId=2939115#l22)

Comment (paulblei) [inline src/main/scala/consul/v1/health/HealthRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759090/issues/source?bid=2788492&fileBranchId=2939115#l22)

File: src/main/scala/consul/v1/agent/check/CheckRequests.scala [hunk 3]
```diff
@@ -46,7 +46,9 @@ trait CheckRequests extends CheckCreators{
 
 object CheckRequests{
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): CheckRequests = new CheckRequests {
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): CheckRequests = new CheckRequests {
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/agent/check/CheckRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759072/issues/source?bid=2788492&fileBranchId=2939115#l51)

Comment (paulblei) [inline src/main/scala/consul/v1/agent/check/CheckRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759072/issues/source?bid=2788492&fileBranchId=2939115#l51)

File: src/main/scala/consul/v1/event/EventRequests.scala [hunk 4]
```diff
@@ -20,7 +20,9 @@ object EventRequests{
 
   private implicit lazy val eventReads = Json.reads[Event]
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext):EventRequests = new EventRequests {
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics):EventRequests = new EventRequests {
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/event/EventRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759086/issues/source?bid=2788492&fileBranchId=2939115#l25)

Comment (paulblei) [inline src/main/scala/consul/v1/event/EventRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759086/issues/source?bid=2788492&fileBranchId=2939115#l25)

File: src/main/scala/consul/v1/agent/service/ServiceRequests.scala [hunk 5]
```diff
@@ -28,7 +28,9 @@ object ServiceRequests {
 
   private implicit lazy val CheckWrites: Writes[Check] = Json.writes[Check]
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): ServiceRequests = new ServiceRequests{
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): ServiceRequests = new ServiceRequests{
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/agent/service/ServiceRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759075/issues/source?bid=2788492&fileBranchId=2939115#l33)

Comment (paulblei) [inline src/main/scala/consul/v1/agent/service/ServiceRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759075/issues/source?bid=2788492&fileBranchId=2939115#l33)

File: src/main/scala/consul/v1/agent/AgentRequests.scala [hunk 6]
```diff
@@ -30,7 +29,9 @@ trait AgentRequests {
 
 object AgentRequests {
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): AgentRequests = new AgentRequests {
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): AgentRequests = new AgentRequests {
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/agent/AgentRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759071/issues/source?bid=2788492&fileBranchId=2939115#l34)

Comment (paulblei) [inline src/main/scala/consul/v1/agent/AgentRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759071/issues/source?bid=2788492&fileBranchId=2939115#l34)

File: src/main/scala/consul/v1/acl/AclRequests.scala [hunk 7]
```diff
@@ -23,7 +23,9 @@ trait AclRequests {
 
 object AclRequests{
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): AclRequests = new AclRequests{
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): AclRequests = new AclRequests{
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/acl/AclRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759068/issues/source?bid=2788492&fileBranchId=2939115#l28)

Comment (paulblei) [inline src/main/scala/consul/v1/acl/AclRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759068/issues/source?bid=2788492&fileBranchId=2939115#l28)

File: src/main/scala/consul/v1/kv/KvRequests.scala [hunk 8]
```diff
@@ -16,7 +16,9 @@ trait KvRequests {
 
 object KvRequests {
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): KvRequests = new KvRequests{
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): KvRequests = new KvRequests{
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/kv/KvRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759092/issues/source?bid=2788492&fileBranchId=2939115#l21)

Comment (paulblei) [inline src/main/scala/consul/v1/kv/KvRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759092/issues/source?bid=2788492&fileBranchId=2939115#l21)

File: src/main/scala/consul/v1/session/SessionRequests.scala [hunk 9]
```diff
@@ -30,7 +30,9 @@ object SessionRequests{
   private lazy implicit val SessionInfoReads     = Json.reads[SessionInfo]
   private lazy implicit val SessionDefWrites     = Json.writes[SessionDef]
 
-  def apply(basePath: String)(implicit executionContext: ExecutionContext): SessionRequests = new SessionRequests{
+  def apply(basePath: String)(implicit executionContext: ExecutionContext, rb: ConsulRequestBasics): SessionRequests = new SessionRequests{
+
+    import rb._
```
Comment (paulblei) [inline src/main/scala/consul/v1/session/SessionRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759094/issues/source?bid=2788492&fileBranchId=2939115#l35)

Comment (paulblei) [inline src/main/scala/consul/v1/session/SessionRequests.scala]: Codacy Issue found: [Put imports in the beginning of the package](https://www.codacy.com/app/Codacy/scala-consul/file/1255759094/issues/source?bid=2788492&fileBranchId=2939115#l35)

File: src/main/scala/consul/v1/common/ConsulRequestBasics.scala [hunk 10]
```diff
@@ -36,15 +39,18 @@ object ConsulRequestBasics {
     }
   }
 
-  private def genRequestMaker[A,B](path: String, httpFunc: WSRequest => Future[WSResponse])(responseTransformer: WSResponse => B)(body: B => A)(implicit executionContext: ExecutionContext): Future[A] = {
-
-    Try(httpFunc(WS.url(path))) match {
+  private def genRequestMaker[A,B](path: String, httpFunc: HttpFunc)(responseTransformer: WSResponse => B)(body: B => A)(implicit executionContext: ExecutionContext): Future[A] = {
+    Try((withToken(token) andThen httpFunc)(WS.url(path))) match {
```
Comment (paulblei) [inline src/main/scala/consul/v1/common/ConsulRequestBasics.scala]: Codacy Issue found: [Default case not specified.](https://www.codacy.com/app/Codacy/scala-consul/file/1255759080/issues/source?bid=2788492&fileBranchId=2939115#l43)

Comment (paulblei) [inline src/main/scala/consul/v1/common/ConsulRequestBasics.scala]: Codacy Issue found: [Default case not specified.](https://www.codacy.com/app/Codacy/scala-consul/file/1255759080/issues/source?bid=2788492&fileBranchId=2939115#l43)

Comment (paulblei) [inline src/main/scala/consul/v1/common/ConsulRequestBasics.scala]: Codacy Issue found: [Default case not specified.](https://www.codacy.com/app/Codacy/scala-consul/file/1256928666/issues/source?bid=2788492&fileBranchId=2939115#l43)

**Comment by rtfpessoa** [comment]
LGTM :+1:
@johannegger can you take a look?

**Pull request merged**